FinalLocalVariable throws a NPE on Switch expression in assignment

Check documentation: https://checkstyle.sourceforge.io/config_coding.html#FinalLocalVariable

Discovered at https://github.com/checkstyle/checkstyle/pull/8920#pullrequestreview-518815743

FinalLocalVariable is throwing a NPE when a switch expression is used on the RHS of an assignment expression:

```
➜  src /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 ExpressionSwitchBugs.java
➜  src cat config.xml                                                                               
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="FinalLocalVariable"/>
    </module>
</module>
➜  src cat ExpressionSwitchBugs.java 
public class ExpressionSwitchBugs {
    private static final int staticValue = 2;
    private static final int staticField = switch(staticValue) {
        case 0 -> -1;
        case 2 -> 2;
        default -> 3;
    };
}
➜  src java -jar /home/nick/IdeaProjects/checkstyle/target/checkstyle-8.37-SNAPSHOT-all.jar -c config.xml ExpressionSwitchBugs.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing ExpressionSwitchBugs.java
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:311)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:221)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:408)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:331)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:190)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:125)
Caused by: java.lang.NullPointerException
        at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.updateUninitializedVariables(FinalLocalVariableCheck.java:482)
        at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.updateAllUninitializedVariables(FinalLocalVariableCheck.java:471)
        at com.puppycrawl.tools.checkstyle.checks.coding.FinalLocalVariableCheck.leaveToken(FinalLocalVariableCheck.java:345)
        at com.puppycrawl.tools.checkstyle.TreeWalker.notifyLeave(TreeWalker.java:357)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:454)
        at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:278)
        at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:151)
        at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:87)
        at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:337)
        at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:298)
        ... 5 more
Checkstyle ends with 1 errors.

```
I would expect FinalLocalVariable to be able to successfully check this file.

--- a/.ci/openjdk14-excluded.files
+++ b/.ci/openjdk14-excluded.files
@@ -8,11 +8,6 @@
   <property name="fileNamePattern" value="[\\/]NestedTypes.java$"/>
 </module>
 
-<!-- until https://github.com/checkstyle/checkstyle/issues/8934 -->
-<module name="BeforeExecutionExclusionFileFilter">
-  <property name="fileNamePattern" value="[\\/]ExpressionSwitchBugs.java$"/>
-</module>
-
 <!-- until https://github.com/checkstyle/checkstyle/issues/8656 -->
 <module name="BeforeExecutionExclusionFileFilter">
   <property name="fileNamePattern" value="[\\/]UnicodeCommentDelimiter.java$"/>

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheck.java
@@ -467,10 +467,13 @@ public class FinalLocalVariableCheck extends AbstractCheck {
      */
     private void updateAllUninitializedVariables(
             Deque<DetailAST> prevScopeUninitializedVariableData) {
-        // Check for only previous scope
-        updateUninitializedVariables(prevScopeUninitializedVariableData);
-        // Check for rest of the scope
-        prevScopeUninitializedVariables.forEach(this::updateUninitializedVariables);
+        final boolean hasSomeScopes = !currentScopeAssignedVariables.isEmpty();
+        if (hasSomeScopes) {
+            // Check for only previous scope
+            updateUninitializedVariables(prevScopeUninitializedVariableData);
+            // Check for rest of the scope
+            prevScopeUninitializedVariables.forEach(this::updateUninitializedVariables);
+        }
     }
 
     /**

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheckTest.java
@@ -301,7 +301,7 @@ public class FinalLocalVariableCheckTest
             "50:19: " + getCheckMessage(MSG_KEY, "e"),
             "88:19: " + getCheckMessage(MSG_KEY, "e"),
             "122:19: " + getCheckMessage(MSG_KEY, "e"),
-            };
+        };
         verify(checkConfig,
             getNonCompilablePath("InputFinalLocalVariableCheckSwitchExpressions.java"),
             expected);

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/FinalLocalVariableCheckTest.java
@@ -307,4 +307,17 @@ public class FinalLocalVariableCheckTest
             expected);
     }
 
+    @Test
+    public void testFinalLocalVariableSwitchAssignment() throws Exception {
+        final DefaultConfiguration checkConfig = createModuleConfig(FinalLocalVariableCheck.class);
+        final String[] expected = {
+            "16:13: " + getCheckMessage(MSG_KEY, "a"),
+            "39:13: " + getCheckMessage(MSG_KEY, "b"),
+            "41:21: " + getCheckMessage(MSG_KEY, "x"),
+        };
+        verify(checkConfig,
+            getNonCompilablePath("InputFinalLocalVariableCheckSwitchAssignment.java"),
+            expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableCheckSwitchAssignment.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/finallocalvariable/InputFinalLocalVariableCheckSwitchAssignment.java
@@ -1,0 +1,66 @@
+//non-compiled with javac: Compilable with Java14
+package com.puppycrawl.tools.checkstyle.checks.coding.finallocalvariable;
+
+/* Config:
+ * default
+ */
+public class InputFinalLocalVariableCheckSwitchAssignment {
+    private static final int staticValue = 2;
+    private static final int staticField = switch(staticValue) {
+        case 0 -> -1;
+        case 2-> 2;
+        default -> 3;
+    };
+
+    public InputFinalLocalVariableCheckSwitchAssignment() throws Exception {
+        int a = 0; // violation
+        final int b = 0;
+        int c = 10;
+        int d = 0;
+
+        c = switch (a) {
+            case 0 -> 5;
+            case 1 -> 10;
+            default -> switch(b) {
+                    case 2 -> {
+                        if (a == 0) {
+                            d = 1; // reassign d
+                        }
+                        throw new Exception();
+                    }
+                    default -> 2;
+                };
+        };
+    }
+
+    public void foo() throws Exception {
+        final int a = 0;
+
+        int b = switch(a) { // violation
+            case 0 -> {
+                int x = 5; // violation
+                int y = 6;
+                if (a == 2) {
+                    y = 7;
+                }
+                throw new Exception();
+            }
+            default -> 2;
+        };
+
+        int c = switch(b) {
+            case 0 -> 1;
+            default -> 2;
+        };
+
+        c = switch(a) {
+            case 0 -> switch (b) {
+                    case 0 -> 1;
+                    case 1 -> 2;
+                    default -> 3;
+                };
+            default -> 1;
+        };
+    }
+
+}

